tests/js/test-utils.js

```diff
@@ -115,6 +115,11 @@ const customRender = ( ui, options = {} ) => {
 		registry,
 		history,
 		waitForRegistry: () => act( waitForRegistry ),
+		waitForRegistryAndTimeouts: () =>
+			act( async () => {
+				await waitForRegistry();
+				return global.waitForTimeouts();
+			} ),
```

Comment:
Since the main problem here is (as you've identified) a [`setTimeout` in the registry itself](https://github.com/WordPress/gutenberg/blob/07baf5a12007d31bbd4ee22113b07952f7eacc26/packages/data/src/namespace-store/index.js#L296-L310), I think it makes sense to make the necessary change here to `waitForRegistry` rather than introducing a new utility. In my initial testing, it seems like all we need to do additionally await a `setTimeout` in that function, i.e.

```diff
diff --git a/tests/js/utils.js b/tests/js/utils.js
index f16ded614c..00e4762e19 100644
--- a/tests/js/utils.js
+++ b/tests/js/utils.js
@@ -503,7 +503,10 @@ export const createWaitForRegistry = ( registry ) => {
 	// We unsubscribe afterwards to allow for potential additions while
 	// Promise.all is resolving.
 	return async () => {
-		await Promise.all( updates );
+		await Promise.all( [
+			...updates,
+			new Promise( ( r ) => setTimeout( r ) ),
+		] );
 		unsubscribe();
 	};
 };
```

This resulted in 9 failed tests (which look potentially legit) when running all `datastore|components` tests and simply reverting the previous use of `waitForRegistry` instead of `waitForRegistryAndTimeouts` in those tests as well as disabling everything in `track-timeouts`.

The above diff could potentially be improved but I think it shows that we could be much closer to a simpler solution here than we thought.

Reply:
I've updated `waitForRegistry` to wait for pending timeouts. I've added a 2ms delay here and elsewhere to avoid a race condition with the default timeouts - see the new [`waitForDefaultTimeouts`](https://github.com/google/site-kit-wp/blob/6f777e9daa57ccaf29455f23387ec76c6a04d290/tests/js/utils.js#L508) utility. Without this, some tests will still fail occasionally. I've documented this in the JSDoc for the `waitForDefaultTimeouts`.

I have been wondering if it should simply be called `waitForTimeouts`, due to the caveat mentioned in the JSDoc, what do you think?

https://github.com/google/site-kit-wp/blob/6f777e9daa57ccaf29455f23387ec76c6a04d290/tests/js/utils.js#L496-L498